src/main/kotlin/org/radarbase/push/integration/garmin/backfill/GarminRequestGenerator.kt

```diff
@@ -71,27 +71,34 @@ class GarminRequestGenerator(
         )
     )
 
-    override fun requests(user: User, max: Int): List<RestRequest> {
-        return routes.map { route ->
-            val offsets: Offsets? = offsetPersistenceFactory.read(user.versionedId)
-            val startDate = userRepository.getBackfillStartDate(user)
-            val startOffset: Instant = if (offsets == null) {
-                logger.debug("No offsets found for $user, using the start date.")
-                startDate
-            } else {
-                logger.debug("Offsets found in persistence.")
-                offsets.offsetsMap.getOrDefault(
-                    UserRoute(user.versionedId, route.toString()), startDate
-                ).takeIf { it >= startDate } ?: startDate
-            }
-            val endDate = userRepository.getBackfillEndDate(user)
-            val endTime = when {
-                endDate <= startOffset -> return@map emptyList() // Already at end. No further requests
-                endDate < startOffset.plus(defaultQueryRange) -> endDate
-                else -> startOffset.plus(defaultQueryRange)
-            }
-            route.generateRequests(user, startOffset, endTime, max / routes.size)
-        }.flatten()
+    private var nextRequestTime: Instant = Instant.MIN
+
+    private val shouldBackoff: Boolean
+        get() = Instant.now() < nextRequestTime
+
+    override fun requests(user: User, max: Int): Sequence<RestRequest> {
+        return if (!shouldBackoff) {
+            routes.map { route ->
+                val offsets: Offsets? = offsetPersistenceFactory.read(user.versionedId)
+                val startDate = userRepository.getBackfillStartDate(user)
+                val startOffset: Instant = if (offsets == null) {
+                    logger.debug("No offsets found for $user, using the start date.")
+                    startDate
+                } else {
+                    logger.debug("Offsets found in persistence.")
+                    offsets.offsetsMap.getOrDefault(
+                        UserRoute(user.versionedId, route.toString()), startDate
+                    ).takeIf { it >= startDate } ?: startDate
+                }
+                val endDate = userRepository.getBackfillEndDate(user)
+                val endTime = when {
+                    endDate <= startOffset -> return@map emptyList() // Already at end. No further requests
+                    endDate < startOffset.plus(defaultQueryRange) -> endDate
+                    else -> startOffset.plus(defaultQueryRange)
+                }
+                route.generateRequests(user, startOffset, endTime, max / routes.size)
+            }.flatten().asSequence().takeWhile { !shouldBackoff }
```

Comment:
Probably more efficient to put the `.asSequence` before the `flatten` operation.